connector/mqtt/vernemq/v2k-brigde/app/MqttClient.js

```diff
@@ -1,68 +1,130 @@
+    this.clientCrt = fs.readFileSync(`${this.config.mqtt.tls.certificate.location}`);
+    this.ca = fs.readFileSync(`${this.config.mqtt.tls.ca.location}`);
+
+    // backPressure
+    this.messageQueue = null;
+    this.currentMessageQueueLenght = 0;
+    this.parallelHandlers = this.config.mqtt.parallelHandlers;
+    this.maxQueLength = this.config.mqtt.maxQueLength;
 
-    this.key = fs.readFileSync(path.join(__dirname, `${this.config.mqtt.hostname}.key`));
-    this.clientCrt = fs.readFileSync(path.join(__dirname, `${this.config.mqtt.hostname}.crt`));
-    this.caCrt = fs.readFileSync(path.join(__dirname, 'ca.crt'));
+    // agent messenger
+    this.agentMessenger = agentMessenger;
   }
 
   init() {
-    const mqttOptions = {
+    this.mqttOptions = {
       username: this.username,
       clientId: this.clientId,
-      host: this.config.mqtt.host,
-      port: this.config.mqtt.port,
-      protocol: 'mqtts',
-      ca: this.caCrt,
-      key: this.key,
+      host: this.host,
+      port: this.port,
+      protocol: this.secureMode ? 'mqtts' : 'mqtt',
+      ca: this.ca,
+      key: this.privateKey,
       cert: this.clientCrt,
-      keepAlive: this.config.mqtt.keepAlive,
+      keepAlive: this.keepAlive,
       clean: false,
-      rejectUnauthorized: true,
+      rejectUnauthorized: false,
     };
 
     const onConnectBind = this.onConnect.bind(this);
     const onDisconnectBind = this.onDisconnect.bind(this);
     const onMessageBind = this.onMessage.bind(this);
 
-    this.mqttc = mqtt.connect(mqttOptions);
+    this.connect();
     this.mqttc.on('connect', onConnectBind);
     this.mqttc.on('disconnect', onDisconnectBind);
     this.mqttc.on('message', onMessageBind);
+
+    /**
+     * Create async queue
+     */
+    this.messageQueue = async.queue((data, done) => {
+      this.asyncQueueWorker(data);
+      done();
+    }, this.parallelHandlers);
+
+    /**
+     * When processing was finalized, reconnect to broker
+     */
+    this.messageQueue.drain(() => {
+      if (this.isConnected === false) {
+        this.mqttc.reconnect();
+      }
+    });
   }
 
   onConnect() {
     this.isConnected = true;
     logger.info(`Client ${this.clientId} connected successfully!`, TAG);
-
-    if (this.agentMessenger === null) {
-      this.agentMessenger = new AgentMessenger(this.config);
-      this.agentMessenger.init(this.mqttc);
-    }
+    this.subscribe();
   }
 
   onDisconnect() {
+    logger.info(`Client ${this.clientId} disconnected, reconnecting ......`);
     this.isConnected = false;
     this.mqttc.reconnect();
   }
 
-  onMessage(topic, message) {
+  onMessage(topic, message, packet) {
+    // pause
+    if (this.isConnected) {
+      // check if message is duplicated
+      if (packet.dup === false) {
+        this.currentMessageQueueLenght += 1;
+        const data = { topic, message };
+        this.messageQueue.push(data, () => {
+          this.currentMessageQueueLenght -= 1;
```

Comment:
It would be more interesting a queue based on bytes size instead of number of messages.